integration/internal/cmdtest/cmd.go

```diff
@@ -0,0 +1,145 @@
+package cmdtest
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+)
+
+type Cmd struct {
+	Path   string
+	Env    []string
+	Args   []string
+	Stdin  io.Reader
+	Stdout io.Writer
+	Stderr io.Writer
+	Silent bool
+}
+
+func (cmd Cmd) WithEnv(env ...string) Cmd {
+	newEnv := make([]string, len(cmd.Env))
+	copy(newEnv, cmd.Env)
+	newEnv = append(newEnv, env...)
+
+	cmd.Env = newEnv
+	return cmd
+}
+
+func (cmd Cmd) WithArgs(args ...string) Cmd {
+	newArgs := make([]string, len(cmd.Args))
+	copy(newArgs, cmd.Args)
+	newArgs = append(newArgs, args...)
+
+	cmd.Args = newArgs
+	return cmd
+}
+
+func (cmd Cmd) WithTempHome(t *testing.T) (Cmd, string) {
+	home := t.TempDir()
+	return cmd.WithEnv("HOME=" + home), home
+}
+
+func (cmd Cmd) Silence() Cmd {
+	cmd.Silent = true
+	return cmd
+}
+
+func (cmd Cmd) WithInput(in io.Reader) Cmd {
+	cmd.Stdin = in
+	return cmd
+}
+
+func (cmd Cmd) OutputTo(out io.Writer) Cmd {
+	cmd.Stdout = out
+	return cmd
+}
+
+func (cmd Cmd) Run(args ...string) error {
```

Comment:
Now I'm debating whether or not it makes sense to return an error here or to assert internally and add a `Cmd.ExpectFailure()` or something that you call prior to the `cmd.{Run,Output,OutputJSON}` that flips the assertion.

I expect that you'll nearly always want the command to succeed, so the extra `NoError` assertions just add clutter (though, I've admittedly gotten pretty used to skimming over `Expect(err).ToNot(HaveOccurred())` and `require.NoError(t, err)`, so perhaps it's not so bad). The `ExpectFailure` API is a bit awkward since you define it before actually running the command, e.g.:

```
fly.ExpectFailure().Run("trigger-job", ...)
```

As a possible alternative, we could have e.g. `Cmd.FallibleRun` that returns an `error` - also awkward (brutal name), and would presumably warrant `FallibleOutput`, etc.

---

What do you think? Perhaps we should just keep things as they are and return errors when we `Run` things - the clutter isn't *so* bad. IMO test helpers shouldn't return errors in general, but making sure a `Cmd.Run` actually succeeds often is the assertion of a test, so maybe it's better for the error checking to be explicit

Reply:
Yeah this is one thing I've been considering, since e.g. `WithTempDir` already calls a testing function which fails the test no any `err` instead of returning it. I can give it a try and see how the code looks!

Reply:
One problem with `ExpectFailure`: sometimes you don't know if you can expect it.

```go
t.Eventually(func() {
  err := fly.Run("login", ...)
  if err != nil {
    return false
  }

  // ...
}, // ...)
```

In this case your second suggestion would probably be better - maybe something like `Try`, `TryOutput`, etc.?

Reply:
In combination with something like this:

```go
cmd.ExpectExit(2).Run(t, "foo")
```

This works nicely since I can just have a struct field initialized to `0` as the expected exit code.

Reply:
Ah good point. And haha Try is much better name. ExpectExit makes a lot of sense too, though I wonder if asserting on the specific exit code could be brittle?

Reply:
It's useful in some cases - for example, we expect `fly execute` to exit 2 for an error or 1 for a failure<sup>[1]</sup>. I think it's reasonable for tests to have strict expectations there in case commands start failing for the "wrong reasons" - e.g. what if the test misconfigured a task which it expected to fail? If it isn't precise enough, this could be an incorrectly passing test.

If you don't care about the exit code I guess you could just use `Try` and assert that an error was returned. I think typically you'd assume 1 and if you get something different back you'd verify that it's correct and then update the test.

[1]: https://github.com/concourse/concourse/blob/6fd0fcbe0fcb529e46e709fece52ba3741318edf/fly/eventstream/render.go#L78-L92

Reply:
Added all this as the last commit - lmk what you think!